### ISSUE TITLE
False-positive on date.setDate() in a loop

### ISSUE THREAD

**Issue by kkomelin** [eslint/eslint#6984]
**What version of ESLint are you using?**

```
"eslint": "^2.10.2",
"eslint-friendly-formatter": "^2.0.5",
"eslint-loader": "^1.3.0",
"eslint-plugin-html": "^1.3.0",
"eslint-config-standard": "^5.1.0",
"eslint-plugin-promise": "^1.0.8",
"eslint-plugin-standard": "^1.3.2",
```

**What parser (default, Babel-ESLint, etc.) are you using?**

"babel-eslint": "^6.1.2",

**Please show your full configuration:**

I am using generated vue-webpack project https://github.com/vuejs-templates/webpack

**What did you do? Please include the actual source code causing the issue.**

```
while (start <= end) { 
      dates.push(new Date(start))
      start.setDate(start.getDate() + 1)
}
```

**What did you expect to happen?**

It should go through eslint checks without any errors.

**What actually happened? Please include the actual, raw output from ESLint.**

>   ✘  http://eslint.org/docs/rules/no-unmodified-loop-condition  'start' is not modified in this loop  
>   ../src/utils/DateUtils.js:163:12
>       while (start <= end) {
>               ^
> 
>   ✘  http://eslint.org/docs/rules/no-unmodified-loop-condition  'end' is not modified in this loop  
>   ../src/utils/DateUtils.js:163:21
>       while (start <= end) {
>                        ^

The solutions suggested:
1. By @ilyavolodin 

```
while (start <= end) {
  dates.push(new Date(start))
  start = start.getDate() + 1;
}
```
1. By @platinumazure 

```
while (start.valueOf() <= end.valueOf())?
```
1. By @platinumazure (my favorite one)

```
while (start <= end) {  // eslint-disable-line no-unmodified-loop-condition
  dates.push(new Date(start))
  start.setDate(start.getDate() + 1)
}
```

**Comment by pmcelhaney** [comment]
I was able to get this to reproduce in the [online demo](http://eslint.org/demo/), but had to declare the variable first.

```js
/*eslint no-unmodified-loop-condition: error*/
var start;
while (start <= end) { 
      dates.push(new Date(start))
      start.setDate(start.getDate() + 1)
}
```

Since we can't know whether calling a function attached to the variable will change its value, I don't think the above case should trigger the error. 

Here's another case where we can't know whether the value of `start` is changed.

```
var start;
while (start <= end) { 
      increment(start);
}
```

**Comment by ilyavolodin** [comment]
@mysticatea Could you provide feedback as to how we might be able to handle those cases?

**Comment by mysticatea** [comment]
`no-unmodified-loop-condition` cannot detect implicit type conversion.

```js
while (start <= end) { 
      dates.push(new Date(start))
      start.setDate(start.getDate() + 1)
}
```

In this case, the variables `start` and `end` are not changed. Those variables have the same instances in the loop. Those are not modified. But `start <= end` triggers implicit type conversion in this case. It's same as `start.valueOf() <= end.valueOf()`. Then, the result of the implicit type conversion is changed in the loop. `no-unmodified-loop-condition` cannot detect the implicit type conversion.

**Comment by mysticatea** [comment]
This is same as #5738.

**Comment by ilyavolodin** [comment]
Should we create an option that would consider call to any functional property on objects as "modified"?

**Comment by platinumazure** [comment]
Some functions are getters, though, and this is somewhat a corner case since it only happens when there are implicit conversions...

I'd prefer something like an array of known setter functions. So if a loop condition variable is the object of a MemberExpression, where the property is a specified setter function, and all of that is the callee of a CallExpression, then we assume the variable is modified.

**Comment by mysticatea** [comment]
My suggestion is "to not use implicit type conversions" or "to disable the rule" here.

For examples: 

- `while (start.getTime() <= end.getTime()) {`
- `while (start <= end) {  //eslint-disable-line no-unmodified-loop-condition`

**Comment by mnpenner** [comment]
Just hit this bug. My loop looks like this:

```js
for(let d = new Date(start); d <= end; d.setDate(d.getDate() + 1)) {
    // ...
}
```

I will disable the inspection for now. Side-question: is there any way we can add a comment beside an `eslint-disable`? e.g.

```js
// eslint-disable-next-line no-unmodified-loop-condition (https://github.com/eslint/eslint/issues/6984)
```

This won't disable the inspection because of the 'comment' I added beside it.

**Comment by kaicataldo** [comment]
I don't think we should allow arbitrary information in the `eslint-disable` comments - having to parse what is config and what isn't would make it error prone. My recommendation would be to put another comment before the `eslint-disable-next-line` comment.

**Comment by kaicataldo** [comment]
I'm closing this issue because it looks like consensus couldn't be reached. While we wish we could accommodate all requests, we have limited resources and do need to prioritize. We've found that issues failing to reach consensus after 21 days tend to never reach consensus, and as a team have decided to close such issues. This doesn't mean the idea isn't interesting or valuable, only that it's not something the team can commit to at the moment.

**Comment by mnpenner** [comment]
I suggest adding an option to `no-unmodified-loop-condition` that allows you to either specify a regex which matches on method names which should be considered to mutate state.

Then we can do something like `mutators: /^set/` and then any method call to a "set" method would be considered to modify the variable. This would fix the issue with dates, but it's *possible* it would also result in a few scenarios where the error isn't caught (false negatives), such as in immutable libraries.

However, I think I'd rather have a few false negatives than to have to suppress them via line-comments all over the place. In either case, making this an option gives the user the choice.

**Comment by kaicataldo** [comment]
@mnpenner Feel free to make a new issue with your proposal

**Comment by feross** [comment]
Just want to note that a `standard` user just ran into this same issue with the following code, using the Moment library:

```js
const dateStart = moment('2016-02-01')
const dateEnd = moment()
while (dateStart < dateEnd) {
  // subtract 1 month
  dateEnd.subtract(1, 'month')
}
```

@mnpenner Want to make a new issue for your proposal?

**Comment by mnpenner** [comment]
@feross @kaicataldo [Here it is](https://github.com/eslint/eslint/issues/8021). Sorry I didn't write it up sooner, been busy. Hopefully that's sufficient :-)
